fastfield_codecs/src/compact_space/mod.rs

```diff
@@ -0,0 +1,726 @@
+            // binary search can never be 0
+            .map_or_else(|e| e - 1, |v| v);
+
+        let range_mapping = &self.ranges_mapping[pos];
+        let diff = compact - range_mapping.compact_start;
+        range_mapping.value_range.start() + diff as u128
+    }
+}
+
+pub struct CompactSpaceCompressor {
+    params: IPCodecParams,
+}
+#[derive(Debug, Clone)]
+pub struct IPCodecParams {
+    compact_space: CompactSpace,
+    bit_unpacker: BitUnpacker,
+    min_value: u128,
+    max_value: u128,
+    num_vals: u64,
+    num_bits: u8,
+}
+
+impl CompactSpaceCompressor {
+    /// Taking the vals as Vec may cost a lot of memory. It is used to sort the vals.
+    pub fn train_from(
+        vals: impl Iterator<Item = u128>,
+        total_num_values_incl_nulls: usize,
+    ) -> Self {
+        let mut tree = BTreeSet::new();
+        tree.extend(vals);
+        assert!(tree.len() <= total_num_values_incl_nulls);
+        train(&tree, total_num_values_incl_nulls)
+    }
+
+    fn write_footer(self, writer: &mut impl Write) -> io::Result<()> {
+        let writer = &mut CountingWriter::wrap(writer);
+        self.params.serialize(writer)?;
+
+        let footer_len = writer.written_bytes() as u32;
+        footer_len.serialize(writer)?;
+
+        Ok(())
+    }
+
+    pub fn compress(self, vals: impl Iterator<Item = Option<u128>>) -> io::Result<Vec<u8>> {
+        let mut output = vec![];
+        self.compress_into(vals, &mut output)?;
+        Ok(output)
+    }
+
+    pub fn compress_into(
+        self,
+        vals: impl Iterator<Item = Option<u128>>,
+        write: &mut impl Write,
+    ) -> io::Result<()> {
+        let mut bitpacker = BitPacker::default();
+        for val in vals {
+            let compact = if let Some(val) = val {
+                self.params.compact_space.to_compact(val).map_err(|_| {
+                    io::Error::new(
+                        io::ErrorKind::InvalidData,
+                        "Could not convert value to compact_space. This is a bug.",
+                    )
+                })?
+            } else {
+                NULL_VALUE_COMPACT_SPACE
+            };
+            bitpacker.write(compact, self.params.num_bits, write)?;
+        }
+        bitpacker.close(write)?;
+        self.write_footer(write)?;
+        Ok(())
+    }
+}
+
+fn train(values_sorted: &BTreeSet<u128>, total_num_values: usize) -> CompactSpaceCompressor {
+    let compact_space = get_compact_space(values_sorted, total_num_values, COST_PER_BLANK_IN_BITS);
+    let amplitude_compact_space = compact_space.amplitude_compact_space();
+
+    assert!(
+        amplitude_compact_space <= u64::MAX as u128,
+        "case unsupported."
+    );
+
+    let num_bits = tantivy_bitpacker::compute_num_bits(amplitude_compact_space as u64);
+    let min_value = *values_sorted.iter().next().unwrap_or(&0);
+    let max_value = *values_sorted.iter().last().unwrap_or(&0);
+    assert!(
```

Comment:
assert_eq

Reply:
The `assert_eq` was not possible with the current algorithm, so I changed it a little bit

The reason it did not work was that we start by collecting `BlankRange`, which implies a minimum size of 3. That's fine for all cases except the borders, since it does not imply a cost by creating a covered range.
So if we had as max value `u128::MAX -1`, we did not add the blank range.